### PR TITLE
Impl create wrapped child tokens, and method to unwrap them

### PR DESCRIPTION
**Pull request by csakoda**

### CODE REVIEW

**Comment by csakoda** [comment]
@greglook

File: src/vault/client.clj [hunk 1]
```diff
@@ -241,7 +289,30 @@
...
+      (when (= (:status response) 200)
+        (:body response))))
+
+  (unwrap!
+    [this wrap-token]
+    (let [response (api-request :post (str api-url "/v1/sys/wrapping/unwrap")
+                                {:headers (build-headers wrap-token)
+                                 :accept :json
+                                 :as :json})]
+      (log/debug "Unwrapping response for" wrap-token)
```
Comment (csakoda) [inline src/vault/client.clj]: Safe to include this context at this point..?

File: src/vault/client.clj [hunk 2]
```diff
@@ -70,7 +79,41 @@
...
+
+  (create-token!
+    [this wrap-ttl]
+    {:request_id "",
+     :lease_id "",
+     :renewable false,
+     :lease_duration 0,
+     :data nil,
+     :wrap_info
+     {:token "87e771a4-73a5-8eda-fe59-a5cc5d41a6d2",
```
Comment (greglook) [inline src/vault/client.clj]: Generate these instead of hard-coding them? You're also ignoring `wrap-ttl` here.

File: src/vault/client.clj [hunk 3]
```diff
@@ -70,7 +79,41 @@
   (delete-secret!
     [this path]
     (swap! memory dissoc path)
-    true))
+    true)
+
+  (create-token! [this] (create-token! this nil))
```
Comment (greglook) [inline src/vault/client.clj]: Format this consistently with the other methods, please.

File: src/vault/client.clj [hunk 4]
```diff
@@ -70,7 +79,41 @@
...
+     {:token "87e771a4-73a5-8eda-fe59-a5cc5d41a6d2",
+      :ttl 300,
+      :creation_time "2016-11-15T16:09:05.570104718-08:00",
+      :wrapped_accessor "2bde7146-78a2-541c-a073-59d6f1535186"},
+     :warnings nil,
+     :auth nil})
+
+  (unwrap!
+    [this wrap-token]
+    {:request_id "21a1a583-a03b-f465-27a9-0461979c7aa7",
```
Comment (greglook) [inline src/vault/client.clj]: I would make this actually look up something from the client's internal state - perhaps placed there by `create-token!`?

File: src/vault/client.clj [hunk 5]
```diff
@@ -167,6 +210,11 @@
                  app (str/join ", " (get-in response [:body :auth :policies])))
       (reset! token-ref client-token))))
 
+(defn- build-headers
+  [token & [wrap-ttl]]
```
Comment (greglook) [inline src/vault/client.clj]: Make this a map argument instead so it's simpler to extend later if needed:
```clojure
[token opts]
```

File: src/vault/client.clj (hunk 1 above)
Comment (greglook) [inline src/vault/client.clj]: I still wouldn't log it. We can add it back later if it turns out to be useful.

**Comment by csakoda** [comment]
@greglook Updated MockClient logic available for your review

File: src/vault/client.clj [hunk 6]
```diff
@@ -241,7 +302,30 @@
                       :as :json})]
       (log/debug "Deleted secret" path)
       (cache/invalidate! cache path)
-      (= (:status response) 204))))
+      (= (:status response) 204)))
+
+  (create-token! [this] (create-token! this nil))
```
Comment (greglook) [inline src/vault/client.clj]: Formatting.

File: src/vault/client.clj [hunk 7]
```diff
@@ -241,7 +302,30 @@
...
+  (create-token!
+    [this wrap-ttl]
+    (check-auth! token)
+    (let [response (api-request :post (str api-url "/v1/auth/token/create")
+                                {:headers (build-headers @token {:wrap-ttl wrap-ttl})
+                                 :accept :json
+                                 :as :json})]
+      (log/debug "Created token" (when wrap-ttl "with X-Vault-Wrap-TTL" wrap-ttl))
+      (when (= (:status response) 200)
+        (:body response))))
```
Comment (greglook) [inline src/vault/client.clj]: In the long run it seems more useful to return some custom type for wrapped values. 🤔

**Pull request merged**